add setting to configure default R interpreter path

sharon-wang opened the pull request:
### Summary

- addresses https://github.com/posit-dev/positron/issues/6207 and https://github.com/posit-dev/positron/issues/6208
- introduces new setting `positron.r.interpreters.default`
    - the default interpreter set with `positron.r.interpreters.default` will be recommended for the workspace with Implicit startup behaviour, as part of runtime startup before discovery takes place
- allow `positron.r.customBinaries` and `positron.r.customRootFolders` to start with aliased home directory `~`
    - relative paths are still ignored, but now we will log that we are ignoring them
    - docs PR: https://github.com/posit-dev/positron-website/pull/55

#### Settings UI

<img width="725" alt="image" src="https://github.com/user-attachments/assets/a4b07d31-5ab1-4d42-a750-f5b563ebed0a" />

#### Settings JSON

```json
    "positron.r.interpreters.default": "~/scratch/4.3-arm64/Resources/bin/R"
```

#### R Language Pack Output

```js
2025-03-07 11:15:50.698 [info] R installation discovered: {
  "usable": true,
  "supported": true,
  "reasonDiscovered": [
    "Found in a location specified via user settings"
  ],
  "reasonRejected": null,
  "binpath": "/Users/sashimi/scratch/4.3-arm64/Resources/bin/R",
  "homepath": "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources",
  "semVersion": {
    "options": {},
    "loose": false,
    "includePrerelease": false,
    "raw": "4.3.3",
    "major": 4,
    "minor": 3,
    "patch": 3,
    "prerelease": [],
    "build": [],
    "version": "4.3.3"
  },
  "version": "4.3.3",
  "arch": "arm64",
  "current": false,
  "orthogonal": true,
  "default": true // <-- this is new as of this PR
}
```

- search for `[recommendedWorkspaceRuntime]` to see if the default interpreter path was recommended for the workspace

### Release Notes

#### New Features

- Default R interpreter in Positron can now be configured via settings (#6207)
- `positron.r.customRootFolders` and `positron.r.customBinaries` paths can now start with tilde `~`

### QA Notes

- the default interpreter should not be started by default -- it should show up as the first option in the interpreter list though!

sharon-wang committed:
add `positron.r.interpreters.default` setting

sharon-wang committed:
add function to retrieve default R interpreter path from settings

sharon-wang committed:
handle tilde paths and filter out non-absolute paths for custom R dirs/binaries

sharon-wang committed:
update wording of `positron.r.interpreters.default` setting description

sharon-wang committed:
recommend default R interpreter for implicit startup if set

sharon-wang committed:
remove redundant untildify and check that default path is absolute

sharon-wang commented on extensions/positron-r/src/runtime-manager.ts:
```diff
@@ -32,8 +35,16 @@ export class RRuntimeManager implements positron.LanguageRuntimeManager {
+		if (defaultInterpreterPath) {
+			if (fs.existsSync(defaultInterpreterPath)) {
+				LOGGER.info(`Recommending R runtime from default interpreter path setting: ${defaultInterpreterPath}`);
+				const inst = new RInstallation(defaultInterpreterPath, undefined, [ReasonDiscovered.user]);
```
perhaps the `ReasonDiscovered.user` should be changed to `ReasonDiscovered.setting`? As the settings may or may not be set by the user themselves (could be a default or set by an admin).

melissa-barca replied:
I agree with the intent, but "setting" by itself is a bit vague,maybe user setting since it is technically a user setting regardless of who configured it?

sharon-wang replied:
sounds good to me! I'ved [changed ReasonDiscovered.user to ReasonDiscovered.userSetting](https://github.com/posit-dev/positron/pull/6621/commits/697f79d0a5077feee9adf8bf00f6f933f294dd14)

melissa-barca commented on extensions/positron-r/src/provider.ts:
```diff
@@ -609,8 +619,17 @@ function userRHeadquarters(): string[] {
+				if (path.isAbsolute(item)) {
+					return true;
+				}
+				LOGGER.info(`R custom binary path ${item} is not absolute...ignoring`);
```
Should we verify here that path exists if not, log an appropriate message?

sharon-wang replied:
Added logging where we do the exists check later in the discovery flow: https://github.com/posit-dev/positron/pull/6621/commits/d5160c5c4799c918c401e2e54bc123e5f360fedc

Added some logging to the `recommendedWorkspaceRuntime()` function as well https://github.com/posit-dev/positron/pull/6621/commits/0a0cd85a0bd76aa648c2e96bc172085f97d30a5a

sharon-wang committed:
add logging when binary file or directory is not found

sharon-wang committed:
change `ReasonDiscovered.user` to `ReasonDiscovered.userSetting`

sharon-wang committed:
add `default` to `RMetadataExtra` and `RInstallation`

sharon-wang committed:
add logging to `recommendedWorkspaceRuntime()`

melissa-barca reviewed (dismissed):
LGTM!

jennybc commented on extensions/positron-r/src/r-installation.ts:
```diff
@@ -48,7 +53,7 @@ export enum ReasonDiscovered {
 	/* eslint-enable @typescript-eslint/naming-convention */
 	adHoc = "adHoc",
-	user = "user",
+	userSetting = "userSetting",
```
I've got this building sense that we should just be saying "setting" instead of "user" (already underway) or "userSetting", because I think most of the recent and ongoing work around this is actually for the workbench context and it's really "adminSetting" most of the time. This is not (necessarily) a concrete ask for this PR, but I wanted to express this. And it's conceivable it has implications for this PR.

This also makes me think about `git config --list --show-origin` and how nice it is for understanding which config file (system, user, repo) is responsible for a particular setting. That's yet another level of not conflating all of this or suggesting that it's all happening at the user level.

sharon-wang replied:
Melissa and I also discussed the naming here: https://github.com/posit-dev/positron/pull/6621#discussion_r1980175178

@melissa-barca do you feel strongly about `userSetting` vs `setting`?

Regardless of if the setting is configured by a user or a Workbench Admin, the setting is configured via User Settings (admin workflow involves merging configured settings into the user's settings). I think we are likely to document these settings as "user settings" even in the Workbench Admin docs?

melissa-barca replied:
I do think User Setting will be more clear than Setting. @sharon-wang  is documenting this in the Workbench Admin Guide as user settings. It does bring up a good point that we should call out this description and the potential for the admin to have configured defaults in the Workbench User Guide. Most likely here: https://docs.posit.co/ide/server-pro/user/positron/guide/settings.html#user-settings

Sharon, do you want me to add that request to your existing workbench doc issue or file a new issue for the Workbench team?

sharon-wang replied:
Ok, I'll keep `userSetting` as-is for now.

I've added a couple extra notes to my existing workbench doc issue to cover those in my docs PR!

jennybc commented on extensions/positron-r/src/r-installation.ts:
```diff
@@ -143,6 +149,12 @@ export class RInstallation {
 		this.current = current;
 		this.reasonDiscovered = reasonDiscovered;
 
+		// Check if the installation is the default R interpreter
```
```suggestion
		// Check if the installation is the default R interpreter for Positron
```

I think this will be helpful to remind anyone who comes later that this does not necessarily have anything to do with the system's current or default version of R. In fact, it's almost the opposite? Because if it's OK to use the system's default version of R, you don't need this setting.

jennybc commented on extensions/positron-r/src/r-installation.ts:
```diff
@@ -338,3 +350,22 @@ function isExcludedInstallation(binpath: string): boolean | undefined {
 }
+
+/**
+ * Get the default R interpreter path specified in the settings.
```
```suggestion
 * Get the default R interpreter path specified in Positron settings.
```

Same rationale as above.

jennybc commented on extensions/positron-r/src/runtime-manager.ts:
```diff
@@ -32,8 +35,20 @@ export class RRuntimeManager implements positron.LanguageRuntimeManager {
+		// If the default interpreter path is set and the path exists on the filesystem,
+		// recommend it with implicit startup behavior.
+		const defaultInterpreterPath = getDefaultInterpreterPath();
+		if (defaultInterpreterPath) {
```
Given the other existence checks, is it possible for this to be `false`? Mostly just curious.

sharon-wang replied:
We don't preset a default interpreter path, so if the user hasn't set this, it will be `undefined`. We won't have done any "does this path actually exist?" checks at this point, just checks for path format like resolving the `~` if any and checking that the path looks like an absolute path.

jennybc commented on extensions/positron-r/src/runtime-manager.ts:
```diff
@@ -32,8 +35,20 @@ export class RRuntimeManager implements positron.LanguageRuntimeManager {
+		const defaultInterpreterPath = getDefaultInterpreterPath();
+		if (defaultInterpreterPath) {
+			if (fs.existsSync(defaultInterpreterPath)) {
+				LOGGER.info(`[recommendedWorkspaceRuntime] Recommending R runtime from default interpreter path setting: ${defaultInterpreterPath}`);
```
We could even give the actual name or key of the setting here, to help someone who is troubleshooting.

jennybc commented:
One overall question I have:

Does this setting even make sense at the workspace level? Feels like it does not, in which case I wonder if we should express/enforce that.

It feels like this setting is something that a user (rarely) or an admin (sometimes?) might set, but I don't see how it's relevant to a specific workspace.

sharon-wang committed:
clarify that the default interpreter path is in Positron settings

it doesn't necessarily have anything to do with the system's current or default version of R

Co-authored-by: Jennifer (Jenny) Bryan <jenny.f.bryan@gmail.com>
Signed-off-by: sharon <sharon-wang@users.noreply.github.com>

sharon-wang committed:
include 'positron.r.interpreters.default' setting key in log text

sharon-wang committed:
fix up settings descriptions

- installation --> binary for more precision
- remove 'Mac' as the paths may not apply to Mac

sharon-wang commented:
@jennybc 
> Does this setting even make sense at the workspace level? Feels like it does not, in which case I wonder if we should express/enforce that.
> 
> It feels like this setting is something that a user (rarely) or an admin (sometimes?) might set, but I don't see how it's relevant to a specific workspace.

Agreed! It makes more sense to machine-scope of all these settings (including the equivalent Python ones). I added a note to the "Polish" section of #6202 to:
- "use settings scope `machine` for all of these settings, as they involve specifying paths which are machine-specific, so they should only be configurable at the User-level, not the Workspace-level"

I'll apply that change in a separate PR as I'd like to do some polish across all settings altogether.
